### PR TITLE
Local Interpolation

### PR DESCRIPTION
**Pull request by sslattery**
The purpose of this PR is to refactor interpolation to make it both easier and to introduce functionality that allows one to use the interpolation code in multiple ways:

1. The `PointSet` has been completely removed in favor of the in-kernel `SplineData`. 
2. The interface for `p2g` and `g2p` is now simpler - just pass a `Kokkos::View` or `Cabana::Slice` or other object with `operator()` that contains the point coordinates and tell it how many points there are. @sfogerty this should make your life easier.
3. This significantly reduces memory footprint - no need to allocate data and carry it around for all those particles anymore or worry about that `reserve` issue for memory
4. For the some users (e.g. @rfbird) you can now not only create `SplineData` in a kernel and do kernel-fused `p2g` and `g2p` with a convenient helper object, it is now very clean to do so with multiple variants. For example, cell and node interpolation can happen at the same time in the same kernel with different orders of spline functions (e.g. quadratic and cubic) by making a `SplineData<double,2,Node>` and `SplineData<double,3,Cell>` and writing loops over their constituent entities.

So basically we get fewer objects, a unified implementation, more generality, and a reduction in memory usage. I'd still like to see if we can come up with a nice in-kernel abstraction for loops over local particle stencils but that will be future work.

### CODE REVIEW

File: src/Cajita_Interpolation.hpp [hunk 1]
```diff
@@ -48,22 +52,29 @@ namespace Cajita
...
   entity.
 
-  \param point_set The point set to use for interpolation.
+  \param points The points over which to perform the interpolation. Will be
+  indexed as (point,dim). The subset of indices in each point's interpolation
+  stencil must be contained within the local grid that will be used for the
+  interpolation
+
+  \param num_point The number of points. This is the size of the first
+  dimension of points.
```
Comment (sfogerty) [inline src/Cajita_Interpolation.hpp]: What is the reason for requiring `num_point` as a param if it can be found from the size of the first dimension of `points`?

Comment (sslattery) [reply src/Cajita_Interpolation.hpp]: good question! so it is the first dimension (e.g. the result of `extent(0)`) if you use a `Kokkos::View` but not if you use a `Cabana::Slice` due to the AoSoA indexing. In addition, a user could write their own container for this as long as it has `operator(p,d)`. Therefore, it made sense to have a consistent `size` parameter.

**Comment by sslattery** [comment]
@rfbird @sfogerty @streeve @kwitaechong any input?
